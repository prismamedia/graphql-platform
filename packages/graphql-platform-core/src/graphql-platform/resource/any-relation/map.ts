import { SuperMapOfNamedObject } from '@prismamedia/graphql-platform-utils';
import { AnyRelation } from '../any-relation';

export class AnyRelationMap extends SuperMapOfNamedObject<AnyRelation> {}
