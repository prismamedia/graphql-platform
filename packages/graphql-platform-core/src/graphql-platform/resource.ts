import {
  FlagConfig,
  fromEntries,
  getFlagValue,
  getPlainObjectKeys,
  GraphQLOperationType,
  isPlainObject,
  loadModuleMap,
  Maybe,
  MaybePromise,
  ModuleMapConfig,
  POJO,
} from '@prismamedia/graphql-platform-utils';
import { EventConfigMap, EventEmitter } from '@prismamedia/ts-async-event-emitter';
import inflector from 'inflection';
import { Memoize } from 'typescript-memoize';
import { AnyBaseContext, AnyGraphQLPlatform, BaseContext, Context, CustomContext, NodeSource } from '../graphql-platform';
import {
  CreateOneOperationArgs,
  CreateOneRawValue,
  DeleteOneOperationArgs,
  Operation,
  OperationConstructor,
  OperationConstructorMap,
  OperationEventMap,
  OperationId,
  OperationType,
  operationTypeMap,
  OperationTypeMapConfig,
  UpdateOneOperationArgs,
  UpdateOneRawValue,
} from './operation';
import {
  AnyFieldConfig,
  AnyRelationConfig,
  Component,
  ComponentMap,
  ComponentSet,
  ComponentValue,
  Field,
  FieldConfig,
  FieldMap,
  FieldSet,
  FieldValue,
  InverseRelationMap,
  InverseRelationSet,
  Relation,
  RelationConfig,
  RelationMap,
  RelationSet,
  RelationValue,
  SerializedComponentValue,
  SerializedFieldValue,
  SerializedRelationValue,
} from './resource/component';
import { InvalidNodeValueError } from './resource/error';
import { ResourceMap } from './resource/map';
import { AnyUniqueFullConfig, Unique, UniqueConfig, UniqueFullConfig, UniqueSet } from './resource/unique';
import { VirtualField, VirtualFieldConfig, VirtualFieldMap, VirtualFieldSet } from './resource/virtual-field';
import {
  InputTypeConstructor,
  InputTypeId,
  inputTypeMap,
  OutputTypeConstructor,
  OutputTypeId,
  outputTypeMap,
  SerializedWhereUniqueInputValue,
  WhereInputValue,
  WhereUniqueInputValue,
} from './type';

export * from './resource/component';
export * from './resource/error';
export * from './resource/graph';
export * from './resource/map';
export * from './resource/set';
export * from './resource/unique';
export * from './resource/virtual-field';

export type NodeValue = { [componentName: string]: ComponentValue };

export type SerializedNodeValue = { [componentName: string]: SerializedComponentValue };

export type MaybeResourceAware<T = any> = T | ((args: { resource: Resource }) => T);

export type MaybeResourceMapAware<T = any> = T | ((args: { resourceMap: ResourceMap }) => T);

export type ResourceFilterValue = WhereInputValue | false;

export enum ResourceHookKind {
  PreCreate = 'PRE_CREATE',
  PostCreate = 'POST_CREATE',
  PreUpdate = 'PRE_UPDATE',
  PostUpdate = 'POST_UPDATE',
  PreDelete = 'PRE_DELETE',
  PostDelete = 'POST_DELETE',
}

export type ResourceHookMetaMap<
  TArgs extends POJO = any,
  TCustomContext extends CustomContext = any,
  TBaseContext extends AnyBaseContext = BaseContext
> = Readonly<{ args: TArgs; context: Context<TCustomContext, TBaseContext>; resource: Resource }>;

export type ResourceHookMap<
  TCustomContext extends CustomContext = any,
  TBaseContext extends AnyBaseContext = BaseContext
> = {
  // Create
  [ResourceHookKind.PreCreate]: {
    metas: ResourceHookMetaMap<CreateOneOperationArgs, TCustomContext, TBaseContext>;
    create: CreateOneRawValue;
  };
  [ResourceHookKind.PostCreate]: Readonly<{
    metas: ResourceHookMetaMap<CreateOneOperationArgs, TCustomContext, TBaseContext>;
    // Contains the whole node value: all the fields and relation's ids
    createdNode: SerializedNodeValue;
  }>;

  // Update
  [ResourceHookKind.PreUpdate]: {
    metas: ResourceHookMetaMap<UpdateOneOperationArgs, TCustomContext, TBaseContext> & {
      toBeUpdatedNodeId: WhereUniqueInputValue;
      toBeUpdatedNode?: NodeSource;
    };
    // @deprecated
    toBeUpdatedNodeId: WhereUniqueInputValue;
    update: UpdateOneRawValue;
  };
  [ResourceHookKind.PostUpdate]: Readonly<{
    metas: ResourceHookMetaMap<UpdateOneOperationArgs, TCustomContext, TBaseContext>;
    // Contains the whole node value: all the fields and relation's ids
    updatedNode: SerializedNodeValue;
  }>;

  // Delete
  [ResourceHookKind.PreDelete]: {
    metas: ResourceHookMetaMap<DeleteOneOperationArgs, TCustomContext, TBaseContext>;
    toBeDeletedNodeId: WhereUniqueInputValue;
  };
  [ResourceHookKind.PostDelete]: Readonly<{
    metas: ResourceHookMetaMap<DeleteOneOperationArgs, TCustomContext, TBaseContext>;
    // Contains the whole node value: all the fields and relation's ids
    deletedNode: SerializedNodeValue;
  }>;
};

export interface ResourceConfig<
  TCustomContext extends CustomContext = {},
  TBaseContext extends AnyBaseContext = BaseContext,
  TUniqueFullConfig extends AnyUniqueFullConfig = UniqueFullConfig,
  TFieldConfig extends AnyFieldConfig = FieldConfig<TCustomContext, TBaseContext>,
  TRelationConfig extends AnyRelationConfig = RelationConfig<TCustomContext, TBaseContext>
> {
  /** Optional, this resource's plural form, default: guessed from the resource's name */
  plural?: Maybe<string>;

  /** Optional, a description */
  description?: Maybe<string>;

  /** Optional, a list of fields */
  fields?: ModuleMapConfig<TFieldConfig>;

  /** Optional, a list of relations */
  relations?: ModuleMapConfig<TRelationConfig>;

  /** Required, the list of the unique constraints. At least one is required as it will be used as this resource's identifier */
  uniques?: Maybe<UniqueConfig<TUniqueFullConfig>[]>;

  /** Optional, either this resource is immutable (= all its components are immutables) or not, default: false */
  immutable?: FlagConfig;

  /** Optional, fine-tune the operations */
  operations?: Maybe<OperationTypeMapConfig>;

  /** Optional, a list of virtual fields (= dynamic fields) */
  virtualFields?: MaybeResourceAware<
    ModuleMapConfig<MaybeResourceAware<VirtualFieldConfig<TCustomContext, TBaseContext>>>
  >;

  /** Optional, a filter applied to all the queries, "false" means returning nothing. */
  filter?: Maybe<
    ResourceFilterValue | ((context: Context<TCustomContext, TBaseContext>) => MaybePromise<Maybe<ResourceFilterValue>>)
  >;

  /** Optional, the behavior of the mutations can be customized by applying hooks */
  hooks?: Maybe<EventConfigMap<ResourceHookMap<TCustomContext, TBaseContext>>>;
}

export type AnyResourceConfig = ResourceConfig<any, any, any, any, any>;

export class Resource<TConfig extends AnyResourceConfig = ResourceConfig> extends EventEmitter<
  ResourceHookMap & OperationEventMap
> {
  public constructor(readonly name: string, readonly config: TConfig, readonly gp: AnyGraphQLPlatform) {
    super();

    const pascalCasedName = inflector.camelize(name, false);
    if (name !== pascalCasedName) {
      throw new Error(`The resource's name "${name}" has to be in "Pascal" case: "${pascalCasedName}".`);
    }

    config.hooks && this.onConfig(config.hooks);
  }

  @Memoize()
  public toString(): string {
    return this.name;
  }

  @Memoize()
  public get plural(): string {
    const plural =
      typeof this.config.plural === 'string' && this.config.plural
        ? this.config.plural
        : inflector.pluralize(this.name);

    const pascalCasedPlural = inflector.camelize(plural, false);
    if (plural !== pascalCasedPlural) {
      throw new Error(`The resource's plural "${plural}" has to be in "Pascal" case: "${pascalCasedPlural}".`);
    }

    if (plural === this.name) {
      throw new Error(
        `The singular "${this.name}" and plural "${plural}" forms of a resource have to be different, you should define it with the "plural" parameter.`,
      );
    }

    return plural;
  }

  @Memoize()
  public get camelCasedName(): string {
    return inflector.camelize(this.name, true);
  }

  @Memoize()
  public get camelCasedPlural(): string {
    return inflector.camelize(this.plural, true);
  }

  @Memoize()
  public get description(): string | undefined {
    return typeof this.config.description === 'string' && this.config.description ? this.config.description : undefined;
  }

  @Memoize()
  public get filter(): (context: Context) => MaybePromise<Maybe<ResourceFilterValue>> {
    const filter = this.config.filter;
    if (filter != null) {
      return typeof filter === 'object' || filter === false ? () => filter : filter;
    }

    return () => undefined;
  }

  @Memoize()
  public getFieldMap(): FieldMap {
    const fieldMap = new FieldMap();
    for (const [name, config] of loadModuleMap(this.config.fields)) {
      fieldMap.set(name, new Field(name, config, this));
    }

    return fieldMap;
  }

  @Memoize()
  public getFieldSet(): FieldSet {
    return new FieldSet(this.getFieldMap().values());
  }

  @Memoize()
  public getRelationMap(): RelationMap {
    const relationMap = new RelationMap();
    for (const [name, config] of loadModuleMap(this.config.relations)) {
      const field = this.getFieldMap().get(name);
      if (field) {
        throw new Error(`The relation "${this}.${name}" cannot have the same name than the field "${field}".`);
      }

      relationMap.set(name, new Relation(name, config, this));
    }

    return relationMap;
  }

  @Memoize()
  public getRelationSet(): RelationSet {
    return new RelationSet(this.getRelationMap().values());
  }

  @Memoize()
  public getComponentMap(): ComponentMap {
    const componentMap = new ComponentMap<Field | Relation>([...this.getFieldMap(), ...this.getRelationMap()]);
    if (componentMap.size === 0) {
      throw new Error(`At least one component (= a field or a relation) has to be defined in the resource "${this}".`);
    }

    return componentMap;
  }

  @Memoize()
  public getComponentSet(): ComponentSet {
    return new ComponentSet(this.getComponentMap().values());
  }

  @Memoize((hookKind: ResourceHookKind) => hookKind)
  public hasPreHook(
    hookKind: ResourceHookKind.PreCreate | ResourceHookKind.PreUpdate | ResourceHookKind.PreDelete,
  ): boolean {
    return (
      this.getEventListenerCount(hookKind) > 0 ||
      (ResourceHookKind.PreDelete !== hookKind &&
        [...this.getComponentSet()].some(component => component.getEventListenerCount(hookKind) > 0))
    );
  }

  @Memoize((hookKind: ResourceHookKind) => hookKind)
  public hasPostHook(
    hookKind: ResourceHookKind.PostCreate | ResourceHookKind.PostUpdate | ResourceHookKind.PostDelete,
  ): boolean {
    return this.getEventListenerCount(hookKind) > 0;
  }

  public assertComponent<TComponent extends Component>(component: TComponent): TComponent {
    if (!this.getComponentSet().has(component)) {
      throw new Error(`The component "${component}" does not belong to the resource "${this}".`);
    }

    return component;
  }

  @Memoize()
  public getUniqueSet(): UniqueSet {
    const uniqueSet = new UniqueSet((this.config.uniques || []).map(config => new Unique(config, this)));

    if (uniqueSet.size === 0) {
      throw new Error(`At least one unique constraint has to be defined in the resource "${this}".`);
    }

    if (!uniqueSet.some(unique => unique.isPublic())) {
      throw new Error(
        `At least one "public" unique constraint (= with all its components "public") has to be defined in the resource "${this}".`,
      );
    }

    return uniqueSet;
  }

  @Memoize()
  public getNonNullableUniqueSet(): UniqueSet {
    return this.getUniqueSet().filter(unique => !unique.isNullable());
  }

  @Memoize()
  public getPublicUniqueSet(): UniqueSet {
    return this.getUniqueSet().filter(unique => unique.isPublic());
  }

  @Memoize()
  public getIdentifier(): Unique {
    const identifier = this.getUniqueSet().first(true);

    for (const component of identifier.componentSet) {
      if (component instanceof Relation && component.getTo() === this) {
        throw new Error(
          `A circular reference has been detected as the "${this}"'s identifier (= the first unique constraint) references itself in "${component.name}".`,
        );
      }
    }

    return identifier;
  }

  @Memoize()
  public getFirstPublicUnique(): Unique {
    return this.getPublicUniqueSet().first(true);
  }

  @Memoize()
  public getInverseRelationMap(): InverseRelationMap {
    const inverseRelationMap = new InverseRelationMap();
    for (const resource of this.gp.getResourceMap().values()) {
      for (const relation of resource.getRelationMap().values()) {
        if (relation.getTo() === this) {
          const inverseRelation = relation.getInverse();

          const component = this.getComponentMap().get(inverseRelation.name);
          if (component) {
            throw new Error(
              `The "${relation}"'s inverse relation cannot have the same name than the component "${component}", you may want to define the "inversedBy" property.`,
            );
          }

          const previousInverseRelation = inverseRelationMap.get(inverseRelation.name);
          if (previousInverseRelation) {
            throw new Error(
              `The "${relation}"'s inverse relation cannot have the same name than the "${previousInverseRelation.getInverse()}"'s inverse relation, you may want to define the "inversedBy" property.`,
            );
          }

          inverseRelationMap.set(inverseRelation.name, inverseRelation);
        }
      }
    }

    return inverseRelationMap;
  }

  @Memoize()
  public getInverseRelationSet(): InverseRelationSet {
    return new InverseRelationSet(this.getInverseRelationMap().values());
  }

  @Memoize()
  public isImmutable(): boolean {
    return getFlagValue(this.config.immutable, false);
  }

  public getOperationConstructorMap<TType extends OperationType>(type: TType): OperationConstructorMap<TType> {
    if (!(type in operationTypeMap && operationTypeMap[type])) {
      throw new Error(
        `The operation type "${type}" does not exist, please choose among "${Object.keys(operationTypeMap).join(
          ', ',
        )}".`,
      );
    }

    return operationTypeMap[type] as any;
  }

  public getOperationConstructor<TType extends OperationType, TId extends OperationId<TType>>(
    type: TType,
    id: TId,
  ): OperationConstructor<TType, TId> {
    const constructorMap = this.getOperationConstructorMap(type);

    if (!(id in constructorMap && constructorMap[id])) {
      throw new Error(
        `The ${type} "${id}" does not exist, please choose among "${Object.keys(constructorMap).join(', ')}".`,
      );
    }

    return constructorMap[id] as any;
  }

  @Memoize((type: OperationType, id: string) => [type, id].join('|'))
  public getOperation<
    TType extends OperationType = OperationType,
    TId extends OperationId<TType> = OperationId<TType>,
    TConstructor extends OperationConstructor<TType, TId> = OperationConstructor<TType, TId>
  >(type: TType, id: TId): InstanceType<TConstructor> {
    const constructor = this.getOperationConstructor(type, id);

    return new constructor(type, id, this);
  }

  public getMutation<TId extends OperationId<GraphQLOperationType.Mutation>>(id: TId) {
    return this.getOperation(GraphQLOperationType.Mutation, id);
  }

  public getQuery<TId extends OperationId<GraphQLOperationType.Query>>(id: TId) {
    return this.getOperation(GraphQLOperationType.Query, id);
  }

  public getOperationTypes(): OperationType[] {
    return getPlainObjectKeys(operationTypeMap);
  }

  public getOperationIds<TType extends OperationType>(type: TType): OperationId<TType>[] {
    const constructorMap = this.getOperationConstructorMap(type);

    return getPlainObjectKeys(constructorMap);
  }

  public getOperations(type?: OperationType): Operation[] {
    const operations: Operation[] = [];

    const types: OperationType[] = type ? [type] : this.getOperationTypes();
    for (const type of types) {
      for (const id of this.getOperationIds(type)) {
        operations.push(this.getOperation(type, id));
      }
    }

    return operations;
  }

  @Memoize((id: string) => id)
  public getInputType<
    TId extends InputTypeId = InputTypeId,
    TConstructor extends InputTypeConstructor<TId> = InputTypeConstructor<TId>
  >(id: TId): InstanceType<TConstructor> {
    if (!(id in inputTypeMap && inputTypeMap[id])) {
      throw new Error(
        `The input type "${id}" does not exist, please choose among "${Object.keys(inputTypeMap).join(', ')}".`,
      );
    }

    return new (inputTypeMap[id] as TConstructor)(id, this);
  }

  @Memoize((id: string) => id)
  public getOutputType<
    TId extends OutputTypeId = OutputTypeId,
    TConstructor extends OutputTypeConstructor<TId> = OutputTypeConstructor<TId>
  >(id: TId): InstanceType<TConstructor> {
    if (!(id in outputTypeMap && outputTypeMap[id])) {
      throw new Error(
        `The output type "${id}" does not exist, please choose among "${Object.keys(outputTypeMap).join(', ')}".`,
      );
    }

    return new (outputTypeMap[id] as TConstructor)(id, this);
  }

  @Memoize()
  public getVirtualFieldMap(): VirtualFieldMap {
    const virtualFieldMap = new VirtualFieldMap();

    if (this.config.virtualFields) {
      for (const [name, config] of loadModuleMap(
        typeof this.config.virtualFields === 'function'
          ? this.config.virtualFields({ resource: this })
          : this.config.virtualFields,
      )) {
        const component = this.getComponentMap().get(name);
        if (component) {
          throw new Error(
            `The virtual field "${this}.${name}" cannot have the same name than the component "${component}".`,
          );
        }

        const inverseRelation = this.getInverseRelationMap().get(name);
        if (inverseRelation) {
          throw new Error(
            `The virtual field "${this}.${name}" cannot have the same name than the "${inverseRelation.getInverse()}"'s inverse relation.`,
          );
        }

        virtualFieldMap.set(
          name,
          new VirtualField(name, typeof config === 'function' ? config({ resource: this }) : config, this),
        );
      }
    }

    return virtualFieldMap;
  }

  @Memoize()
  public getVirtualFieldSet(): VirtualFieldSet {
    return new VirtualFieldSet(this.getVirtualFieldMap().values());
  }

  public assertValue(node: unknown, normalized?: boolean, componentSet?: ComponentSet): NodeValue {
    if (isPlainObject(node)) {
      const strict = typeof componentSet !== 'undefined';

      return fromEntries(
        [...(componentSet || this.getComponentSet())].map(component => [
          component.name,
          typeof node[component.name] !== 'undefined' || strict
            ? component.isField()
              ? component.assertValue(node[component.name])
              : component.assertValue(node[component.name], normalized)
            : undefined,
        ]),
      );
    }

    throw new InvalidNodeValueError(this, `a plain object is expected but received "${node}" instead`);
  }

  public serializeValue(node: NodeValue, normalized?: boolean, componentSet?: ComponentSet): SerializedNodeValue {
    const strict = typeof componentSet !== 'undefined';

    return fromEntries(
      [...(componentSet || this.getComponentSet())].map(component => [
        component.name,
        typeof node[component.name] !== 'undefined' || strict
          ? component.isField()
            ? component.serializeValue(node[component.name] as FieldValue)
            : component.serializeValue(node[component.name] as RelationValue, normalized)
          : undefined,
      ]),
    );
  }

  public parseValue(node: SerializedNodeValue, normalized?: boolean, componentSet?: ComponentSet): NodeValue {
    const strict = typeof componentSet !== 'undefined';

    return fromEntries(
      [...(componentSet || this.getComponentSet())].map(component => [
        component.name,
        typeof node[component.name] !== 'undefined' || strict
          ? component.isField()
            ? component.parseValue(node[component.name] as SerializedFieldValue)
            : component.parseValue(node[component.name] as SerializedRelationValue, normalized)
          : undefined,
      ]),
    );
  }

  public assertId(id: unknown): WhereUniqueInputValue {
    return this.getInputType('WhereUnique').assert(id);
  }

  public serializeId(id: WhereUniqueInputValue): SerializedWhereUniqueInputValue {
    return this.serializeValue(this.assertId(id));
  }

  public parseId(node: SerializedWhereUniqueInputValue): WhereUniqueInputValue {
    return this.assertId(this.parseValue(node));
  }
}

export type AnyResource = Resource<any>;
