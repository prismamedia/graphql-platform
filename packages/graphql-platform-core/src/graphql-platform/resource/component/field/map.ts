import { Field } from '../field';
import { ComponentMap } from '../map';

export class FieldMap extends ComponentMap<Field> {}
