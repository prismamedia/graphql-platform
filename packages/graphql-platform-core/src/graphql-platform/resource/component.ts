import { Field } from './component/field';
import { Relation } from './component/relation';
import { Component } from './component/types';

export * from './component/field';
export * from './component/map';
export * from './component/relation';
export * from './component/set';
export * from './component/types';

export function isComponent(maybeComponent: unknown): maybeComponent is Component {
  return maybeComponent && (maybeComponent instanceof Field || maybeComponent instanceof Relation);
}
