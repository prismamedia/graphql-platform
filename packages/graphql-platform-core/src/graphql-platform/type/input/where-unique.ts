import {
  GraphQLNonNullDecorator,
  isPlainObject,
  MaybeUndefinedDecorator,
  SuperMap,
} from '@prismamedia/graphql-platform-utils';
import { GraphQLBoolean, GraphQLInputFieldConfigMap, GraphQLInputObjectType } from 'graphql';
import { Memoize } from 'typescript-memoize';
import { ComponentSet, Field, Unique } from '../../resource';
import { FieldValue, Relation } from '../../resource/component';
import { UniqueSet } from '../../resource/unique';
import { AbstractInputType } from '../abstract-type';

// Node's identifier
export type WhereUniqueInputValue = {
  [componentName: string]: FieldValue | (null | WhereUniqueInputValue);
};

export class WhereUniqueInputType extends AbstractInputType {
  @Memoize()
  public get name() {
    return `${this.resource.name}WhereUniqueInput`;
  }

  @Memoize()
  public get description() {
    return `Identifies exactly one "${this.resource}" node`;
  }

  @Memoize((knownRelation?: Relation, forced: boolean = false) =>
    [knownRelation ? knownRelation.name : '', String(forced)].join('|'),
  )
  public getUniqueCombinationMap(knownRelation?: Relation, forced: boolean = false) {
    // Contains all the resource's unique constraints
    const resourceUniqueSet = this.resource.getUniqueSet();

    if (knownRelation && !resourceUniqueSet.getComponentSet().has(knownRelation)) {
      throw new Error(
        `The relation "${knownRelation}" is not part of the "${
          this.resource
        }"'s unique constraints: ${resourceUniqueSet.getNames().join(', ')}`,
      );
    }

    const uniqueSet =
      knownRelation && forced === true
        ? resourceUniqueSet.filter(({ componentSet }) => componentSet.has(knownRelation))
        : resourceUniqueSet;

    const componentInUniqueSet =
      knownRelation && forced === true ? uniqueSet.getComponentSet().diff(knownRelation) : uniqueSet.getComponentSet();

    const uniqueCombinationMap = new SuperMap<string, ComponentSet>();
    for (const unique of uniqueSet) {
      const displayedComponentInUniqueSet = unique.componentSet.intersect(componentInUniqueSet).sortByName();
      const displayedComponentInUniqueSetKey = displayedComponentInUniqueSet.getNames().join('|');

      if (displayedComponentInUniqueSet.size > 0 && !uniqueCombinationMap.has(displayedComponentInUniqueSetKey)) {
        uniqueCombinationMap.set(displayedComponentInUniqueSetKey, displayedComponentInUniqueSet);
      }
    }

    return uniqueCombinationMap;
  }

  public parseUnique<TStrict extends boolean>(
    value: unknown,
    unique: Unique,
    strict: TStrict,
    // Accept only the defined unique
    relationToUniqueOnly: boolean = false,
  ): MaybeUndefinedDecorator<WhereUniqueInputValue, TStrict> {
    const id: WhereUniqueInputValue = Object.create(null);

    if (
      isPlainObject(value) &&
      unique.componentSet.every(component => {
        if (component.isField()) {
          // Field
          const componentValue = component.getValue(value, false);
          if (typeof componentValue === 'undefined') {
            return false;
          }

          component.setValue(id, componentValue);
        } else {
          // Relation
          const relatedNodeId = component.getId(value, false, relationToUniqueOnly);
          if (typeof relatedNodeId === 'undefined') {
            return false;
          }

          component.setValue(id, relatedNodeId);
        }

        return true;
      })
    ) {
      return id as any;
    }

    if (strict) {
      throw new Error(
        `The following "${this.resource}"'s identifier does not contain a valid "${
          unique.name
        }" combination: ${JSON.stringify(value)}`,
      );
    }

    return undefined as any;
  }

  public parse<TStrict extends boolean>(
    value: unknown,
    strict: TStrict,
    uniqueSet: UniqueSet = this.resource.getUniqueSet(),
  ): MaybeUndefinedDecorator<WhereUniqueInputValue, TStrict> {
    let id: WhereUniqueInputValue | undefined;

    if (isPlainObject(value) && uniqueSet.some(unique => !!(id = this.parseUnique(value, unique, false)))) {
      return id as any;
    }

    if (strict) {
      throw new Error(
        `The following "${this.resource}"'s identifier does not contain any valid unique combination: ${JSON.stringify(
          value,
        )}`,
      );
    }

    return undefined as any;
  }

  @Memoize((knownRelation?: Relation, forced: boolean = true) =>
    [knownRelation ? knownRelation.name : '', String(forced)].join('|'),
  )
  public getGraphQLType(knownRelation?: Relation, forced: boolean = true) {
    const uniqueCombinationMap = this.getUniqueCombinationMap(knownRelation, forced);
    const displayedComponentSet = new ComponentSet().concat(...[...uniqueCombinationMap.values()]);

    return displayedComponentSet.size > 0
      ? new GraphQLInputObjectType({
          name: [
            this.resource.name,
            knownRelation ? `With${forced ? 'Forced' : 'Optional'}${knownRelation.pascalCasedName}` : null,
            'WhereUniqueInput',
          ]
            .filter(Boolean)
            .join(''),
          description: `${
            knownRelation ? `Given a ${forced ? 'forced' : 'known'} "${knownRelation.name}", i` : 'I'
          }dentifies exactly one "${this.resource.name}" node${
            uniqueCombinationMap.size > 1
              ? ` with one of these unique ${
                  uniqueCombinationMap.some(([, componentSet]) => componentSet.size > 1)
                    ? 'combinations of values'
                    : 'values'
                }:\n${[...uniqueCombinationMap.values()]
                  .map(
                    componentSet =>
                      `- ${[...componentSet]
                        .map(component => `${component.name}${component === knownRelation ? ' (optional)' : ''}`)
                        .join(', ')}`,
                  )
                  .join('\n')}`
              : '.'
          }`,
          fields: () => {
            const fields: GraphQLInputFieldConfigMap = {};

            for (const component of displayedComponentSet) {
              fields[component.name] = {
                type: GraphQLNonNullDecorator(
                  component instanceof Field
                    ? component.getType()
                    : component
                        .getTo()
                        .getInputType('WhereUnique')
                        .getGraphQLType(),
                  uniqueCombinationMap.every(
                    ([, componentSet]) => componentSet.has(component) && component !== knownRelation,
                  ),
                ),
              };
            }

            return fields;
          },
        })
      : GraphQLBoolean;
  }
}
