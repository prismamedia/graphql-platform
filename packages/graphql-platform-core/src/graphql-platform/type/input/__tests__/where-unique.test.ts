import { POJO } from '@prismamedia/graphql-platform-utils';
import { printType } from 'graphql';
import { config, MyGP } from '../../../../__tests__/gp';
import { Resource } from '../../../resource';
import { Relation } from '../../../resource/component';
import { ResourceMap } from '../../../resource/map';
import { WhereUniqueInputValue } from '../where-unique';

describe('WhereUniqueInput', () => {
  let gp: MyGP;
  let resourceMap: ResourceMap;

  beforeAll(() => {
    gp = new MyGP(config);
    resourceMap = gp.getResourceMap();
  });

  it.each([
    ['Article', ['category']],
    ['ArticleTag', ['article', 'tag']],
  ] as ReadonlyArray<[Resource['name'], Relation['name'][] | undefined]>)(
    'creates a valid GraphQL "WhereUniqueInput" type',
    (resourceName: Resource['name'], relationNames?: Relation['name'][]) => {
      const resource = resourceMap.assert(resourceName);

      expect(
        printType(resource.getInputType('WhereUnique').getGraphQLType(), {
          commentDescriptions: true,
        }),
      ).toMatchSnapshot();

      if (relationNames) {
        relationNames.forEach((relationName) => {
          const relation = resource.getRelationMap().assert(relationName);

          expect(
            printType(
              resource
                .getInputType('WhereUnique')
                .getGraphQLType(relation, false),
              {
                commentDescriptions: true,
              },
            ),
          ).toMatchSnapshot();

          expect(
            printType(
              resource
                .getInputType('WhereUnique')
                .getGraphQLType(relation, true),
              {
                commentDescriptions: true,
              },
            ),
          ).toMatchSnapshot();
        });
      }
    },
  );

  it('throws an error on unknown component', () => {
    expect(() =>
      resourceMap
        .assert('Article')
        .getInputType('WhereUnique')
        .getGraphQLType(
          resourceMap.assert('Article').getRelationMap().assert('author'),
          false,
        ),
    ).toThrowErrorMatchingInlineSnapshot(
      `"The relation \\"Article.author\\" is not part of the \\"Article\\"'s unique constraints: _id, id, category-slug"`,
    );
  });

  it.each([[undefined], [null], [{}]] as ReadonlyArray<any>)(
    'throws error on invalid values',
    (value) => {
      expect(() =>
        resourceMap.assert('Article').getInputType('WhereUnique').assert(value),
      ).toThrowError(
        'The following "Article"\'s identifier does not contain any valid',
      );
    },
  );

  it.each([
    [
      {
        _id: 100,
        id: '100',
        category: {
          _id: 10,
        },
        slug: 'a-valid-article-slug',
      },
      { _id: 100 },
    ],
    [
      {
        id: '100',
        category: {
          _id: 10,
        },
        slug: 'a-valid-article-slug',
      },
      { id: '100' },
    ],
    [
      {
        category: {
          _id: 10,
          parent: {
            id: '31cc2a48-551a-4b59-82cd-c4336a217dca',
          },
          slug: 'a-valid-category-slug',
        },
        slug: 'a-valid-article-slug',
      },
      {
        category: {
          parent: {
            id: '31cc2a48-551a-4b59-82cd-c4336a217dca',
          },
          slug: 'a-valid-category-slug',
        },
        slug: 'a-valid-article-slug',
      },
    ],
    [
      {
        category: {
          _id: 10,
          parent: null,
          slug: 'a-valid-category-slug',
        },
        slug: 'a-valid-article-slug',
      },
      {
        category: {
          parent: null,
          slug: 'a-valid-category-slug',
        },
        slug: 'a-valid-article-slug',
      },
    ],
  ] as ReadonlyArray<[WhereUniqueInputValue, POJO]>)(
    'parses value into "whereUniqueInputValue"',
    (value, whereUniqueInputValue) => {
      expect(
        resourceMap.assert('Article').getInputType('WhereUnique').parse(value),
      ).toEqual(whereUniqueInputValue);
    },
  );
});
