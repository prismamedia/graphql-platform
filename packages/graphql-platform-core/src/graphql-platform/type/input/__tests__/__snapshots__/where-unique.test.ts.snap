// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WhereUniqueInput creates a valid GraphQL "WhereUniqueInput" type 1`] = `
"# Identifies exactly one \\"Article\\" node with one of these unique combinations of values:
# - _id
# - id
# - category, slug
input ArticleWhereUniqueInput {
  _id: Int
  id: ID
  category: CategoryWhereUniqueInput
  slug: String
}"
`;

exports[`WhereUniqueInput creates a valid GraphQL "WhereUniqueInput" type 2`] = `
"# Given a known \\"category\\", identifies exactly one \\"Article\\" node with one of these unique combinations of values:
# - _id
# - id
# - category (optional), slug
input ArticleWithOptionalCategoryWhereUniqueInput {
  _id: Int
  id: ID
  category: CategoryWhereUniqueInput
  slug: String
}"
`;

exports[`WhereUniqueInput creates a valid GraphQL "WhereUniqueInput" type 3`] = `
"# Given a forced \\"category\\", identifies exactly one \\"Article\\" node.
input ArticleWithForcedCategoryWhereUniqueInput {
  slug: String!
}"
`;

exports[`WhereUniqueInput creates a valid GraphQL "WhereUniqueInput" type 4`] = `
"# Identifies exactly one \\"ArticleTag\\" node with one of these unique combinations of values:
# - article, tag
# - article, order
input ArticleTagWhereUniqueInput {
  article: ArticleWhereUniqueInput!
  tag: TagWhereUniqueInput
  order: Int
}"
`;

exports[`WhereUniqueInput creates a valid GraphQL "WhereUniqueInput" type 5`] = `
"# Given a known \\"article\\", identifies exactly one \\"ArticleTag\\" node with one of these unique combinations of values:
# - article (optional), tag
# - article (optional), order
input ArticleTagWithOptionalArticleWhereUniqueInput {
  article: ArticleWhereUniqueInput
  tag: TagWhereUniqueInput
  order: Int
}"
`;

exports[`WhereUniqueInput creates a valid GraphQL "WhereUniqueInput" type 6`] = `
"# Given a forced \\"article\\", identifies exactly one \\"ArticleTag\\" node with one of these unique values:
# - tag
# - order
input ArticleTagWithForcedArticleWhereUniqueInput {
  tag: TagWhereUniqueInput
  order: Int
}"
`;

exports[`WhereUniqueInput creates a valid GraphQL "WhereUniqueInput" type 7`] = `
"# Given a known \\"tag\\", identifies exactly one \\"ArticleTag\\" node with one of these unique combinations of values:
# - article, tag (optional)
# - article, order
input ArticleTagWithOptionalTagWhereUniqueInput {
  article: ArticleWhereUniqueInput!
  tag: TagWhereUniqueInput
  order: Int
}"
`;

exports[`WhereUniqueInput creates a valid GraphQL "WhereUniqueInput" type 8`] = `
"# Given a forced \\"tag\\", identifies exactly one \\"ArticleTag\\" node.
input ArticleTagWithForcedTagWhereUniqueInput {
  article: ArticleWhereUniqueInput!
}"
`;
